### DISCUSSION TITLE
Does DP(dara_parallel_backend) work in Brain.evaluate() ?

### DISCUSSION THREAD

**Discussion by My-captain** [speechbrain/speechbrain#1922] [Q&A] [answered]
I want to evaluate some pretrained models on my specified datasets fastly, and I have 8 GPUS.
However，DP doesn't seem to work properly，when I run the following command：
`CUDA_VISIBLE_DEVICES=0,1,2,3 nohup python train_with_whisper.py hparams/vannilla_whisper.yaml --data_parallel_backend > medium_train.log &`

I noticed that `on_fit_start` is called in the `fit()`, and `_wrap_distributed` is included, but `evaluate` does not.

Any suggestions please?

**Comment by asumagic** [comment, marked as answer]
FWIW the readthedocs documentation is outdated, so note that DP is deprecated in favor of DDP in SpeechBrain: https://github.com/speechbrain/speechbrain/blob/develop/docs/multigpu.md

If it isn't too difficult for you to migrate to DDP then it's probably less buggy in general.

**Reply by My-captain** [reply]
Got it，Thanks

**Reply by JuanPZuluaga** [reply]
Hello @My-captain, did you manage to solve that issue?

**Reply by My-captain** [reply]
@JuanPZuluaga In order to save time, I just use single GPU for evaluation.

**Reply by Adel-Moumen** [reply]
Hey, the readthedocs documentation is not outdated anymore. You can have a look here: https://speechbrain.readthedocs.io/en/latest/multigpu.html

**Comment by JuanPZuluaga** [comment]
Hello, I'd like to follow up with this question. 

@AsuMagic, it looks like, even with DDP, brain.evaluate() does not take advantage of multi-gpu setup. 

Actually, I'm having quite weird results when I finish training and move on to evaluation with DDP. TL;DR I get worse results while `brain.evaluate` with DDP. 

I realized that in the forward pass, and particularly the part where the outputs of the model (at evaluation stage) are appended to the `wer object (self.wer_metric.append)` ([example](https://github.com/speechbrain/speechbrain/blob/e60037c9c29b80bfbcd873828d25aa16d147c8c0/recipes/LibriSpeech/ASR/transformer/train_with_whisper.py#L105)): this object gets appended the same results multiple times from different GPUs. Because of the system is not aware of multiple gpus running in parallel at evaluation stage, all the GPUs perform evaluation on the whole test set and as `sb.utils.distributed.if_main_process()` is not getting call in the forward `compute_objectives`, the wer object gets several times updated with the same results.

Do you think is there an issue here? 

(Tagging also @TParcollet 👍 and @Adel-Moumen )


Pablo.

**Reply by Adel-Moumen** [reply]
Hey @JuanPZuluaga, 

Thanks for pointing out this. Could you please open an issue directly about this problem? 

BTW, if you have a solution against this issue, please let me know. :-) 

Thanks again guys!
